### PR TITLE
[BUGFIX beta] Fix willRender and willInsertElement ordering and element availability

### PR DESCRIPTION
**Pull request by chadhietala**
This fixes the updating semantics for the `willRender` and `willInsterElement` hooks on a component. Prior to this commit this were not getting called at the correct time.

### CODE REVIEW

File: packages/ember-glimmer/lib/syntax/curly-component.js [hunk 1]
```diff
@@ -210,11 +210,14 @@ class CurlyComponentManager {
 
     // We usually do this in the `didCreateElement`, but that hook doesn't fire for tagless components
     if (component.tagName === '') {
+      if (environment.isInteractive) {
+        component.trigger('willRender');
+      }
```
Comment (chancancode) [inline packages/ember-glimmer/lib/syntax/curly-component.js]: I think you need to move this into `create`? I thought the point is you want to `set` on the component before we apply the bindings

File: packages/ember-glimmer/lib/syntax/curly-component.js [hunk 2]
```diff
@@ -259,6 +262,10 @@ class CurlyComponentManager {
   }
 
   didCreateElement({ component, classRef, environment }, element, operations) {
+    if (environment.isInteractive) {
+      component.trigger('willRender');
+    }
```
Comment (chancancode) [inline packages/ember-glimmer/lib/syntax/curly-component.js]: Move this up to `create` and remove the `tagName` guard

File: packages/ember-glimmer/tests/integration/components/life-cycle-test.js [hunk 3]
```diff
@@ -1378,8 +1381,64 @@ moduleFor('Run loop and lifecycle hooks', class extends RenderingTest {
...
   }
+
+  ['@test element availability in `willInsertElement`'](assert) {
+    let ComponentClass = Component.extend({
+      tagName: 'a',
+      customHref: 'http://google.com',
+      attributeBindings: ['customHref:href'],
+      willInsertElement() {
+        assert.ok(this.element, 'should have element');
+        assert.notOk(this.element.parentNode, 'is detached');
```
Comment (chancancode) [inline packages/ember-glimmer/tests/integration/components/life-cycle-test.js]: Isn't this already covered in the lifecycle tests?

File: packages/ember-glimmer/tests/integration/components/life-cycle-test.js [hunk 4]
```diff
@@ -1378,8 +1381,64 @@ moduleFor('Run loop and lifecycle hooks', class extends RenderingTest {
...
+        id: 'foo',
+        href: 'http://willRender.com',
+        class: classes('ember-view')
+      }
+    });
+  }
 });
 
+
+
```
Comment (chancancode) [inline packages/ember-glimmer/tests/integration/components/life-cycle-test.js]: ✂️

File: packages/ember-glimmer/tests/integration/components/life-cycle-test.js [hunk 5]
```diff
@@ -1378,8 +1381,64 @@ moduleFor('Run loop and lifecycle hooks', class extends RenderingTest {
...
+    });
+  }
+
+  ['@test `willRender` can set before render'](assert) {
+    let ComponentClass = Component.extend({
+      tagName: 'a',
+      customHref: 'http://google.com',
+      attributeBindings: ['customHref:href'],
+      willRender() {
+        assert.notOk(this.element, 'should not have element');
```
Comment (chancancode) [inline packages/ember-glimmer/tests/integration/components/life-cycle-test.js]: Probably ✂️ this assertion ? It's already covered by the general lifecycle tests, so this just distracts from the point of the test imo

File: packages/ember-glimmer/tests/integration/components/life-cycle-test.js [hunk 6]
```diff
@@ -1378,8 +1381,64 @@ moduleFor('Run loop and lifecycle hooks', class extends RenderingTest {
...
+      tagName: 'a',
+      attrs: {
+        id: 'foo',
+        href: 'http://google.com',
+        class: classes('ember-view')
+      }
+    });
+  }
+
+  ['@test `willRender` can set before render'](assert) {
```
Comment (chancancode) [inline packages/ember-glimmer/tests/integration/components/life-cycle-test.js]: Could use a comment, or link to github issue

File: packages/ember-glimmer/lib/syntax/curly-component.js (hunk 2 above)
Comment (chadhietala) [inline packages/ember-glimmer/lib/syntax/curly-component.js]: I actually don't know if we want to move this.

File: packages/ember-glimmer/tests/integration/components/life-cycle-test.js (hunk 3 above)
Comment (chadhietala) [inline packages/ember-glimmer/tests/integration/components/life-cycle-test.js]: Removed.

File: packages/ember-glimmer/tests/integration/components/life-cycle-test.js (hunk 5 above)
Comment (chadhietala) [inline packages/ember-glimmer/tests/integration/components/life-cycle-test.js]: Removed

File: packages/ember-glimmer/lib/syntax/curly-component.js (hunk 2 above)
Comment (chancancode) [inline packages/ember-glimmer/lib/syntax/curly-component.js]: I think the difference is probably not observable. All that would have happened between the end of create and here is that we would have compiled the layout, entered it, created an empty element for in the VM (but haven't appended it into DOM). So it should be okay to align this particular book between tagless and regular components.

Comment (chancancode) [inline packages/ember-glimmer/lib/syntax/curly-component.js]: Do you have other concerns? I could totally have missed something

**Comment by chadhietala** [comment]
Fixes https://github.com/emberjs/ember.js/issues/14458

File: packages/ember-glimmer/lib/syntax/curly-component.js (hunk 2 above)
Comment (chadhietala) [inline packages/ember-glimmer/lib/syntax/curly-component.js]: Actually did some more investigation. Moved this and updated the tests accordingly.

File: packages/ember-glimmer/tests/integration/components/life-cycle-test.js (hunk 6 above)
Comment (chadhietala) [inline packages/ember-glimmer/tests/integration/components/life-cycle-test.js]: Done

**Pull request merged**